cloudbridge/cloud/providers/azure/services.py

```diff
@@ -0,0 +1,1024 @@
+            return True
+        except CloudError as cloudError:
+            # Azure raises the cloud error if the resource not available
+            log.exception(cloudError.message)
+            return False
+
+
+class AzureKeyPairService(BaseKeyPairService):
+    PARTITION_KEY = '00000000-0000-0000-0000-000000000000'
+
+    def __init__(self, provider):
+        super(AzureKeyPairService, self).__init__(provider)
+
+    def get(self, key_pair_id):
+        try:
+            key_pair = self.provider.azure_client.\
+                get_public_key(key_pair_id)
+
+            if key_pair:
+                return AzureKeyPair(self.provider, key_pair)
+            return None
+        except AzureException as error:
+            log.exception(error)
+            return None
+
+    def list(self, limit=None, marker=None):
+        key_pairs = [AzureKeyPair(self.provider, key_pair) for key_pair in
+                     self.provider.azure_client.
+                     list_public_keys(AzureKeyPairService.PARTITION_KEY)]
+        return ClientPagedResultList(self.provider, key_pairs, limit, marker)
+
+    def find(self, name, limit=None, marker=None):
+        key_pair = self.get(name)
+        return ClientPagedResultList(self.provider,
+                                     [key_pair] if key_pair else [],
+                                     limit, marker)
+
+    def create(self, name):
+
+        key_pair = self.get(name)
+
+        if key_pair:
+            raise Exception(
+                'Keypair already exists with name {0}'.format(name))
+
+        private_key_str, public_key_str = azure_helpers.gen_key_pair()
+
+        entity = {
+                  'PartitionKey': AzureKeyPairService.PARTITION_KEY,
+                  'RowKey': str(uuid.uuid4()),
+                  'Name': name,
+                  'Key': public_key_str
+                 }
+
+        self.provider.azure_client.create_public_key(entity)
+
+        key_pair = self.get(name)
+
+        key_pair.material = private_key_str
+
+        return key_pair
+
+
+class AzureObjectStoreService(BaseObjectStoreService):
+    def __init__(self, provider):
+        super(AzureObjectStoreService, self).__init__(provider)
+
+    def get(self, bucket_id):
+        """
+        Returns a bucket given its ID. Returns ``None`` if the bucket
+        does not exist.
+        """
+        try:
+            bucket = self.provider.azure_client.get_container(bucket_id)
+            return AzureBucket(self.provider, bucket)
+
+        except AzureException as error:
+            log.exception(error)
+            return None
+
+    def find(self, name, limit=None, marker=None):
+        """
+        Searches for a bucket by a given list of attributes.
+        """
+        buckets = [AzureBucket(self.provider, bucket)
+                   for bucket in
+                   self.provider.azure_client.list_containers(prefix=name)]
+        return ClientPagedResultList(self.provider, buckets,
```

Comment:
@VikramAIS @madhugilla Also here. I think it may be a good idea to search for all places where a ClientPagedResultList is being used and check whether it's possible to replace it with a ServerPagedResultList.